applylib/applyset/applyset.go

```diff
@@ -153,11 +242,151 @@ func (a *ApplySet) ApplyOnce(ctx context.Context) (*ApplyResults, error) {
 			results.applyError(gvk, nn, fmt.Errorf("error from apply: %w", err))
 			continue
 		}
-
+		visitedUids.Insert(lastApplied.GetUID())
 		tracker.lastApplied = lastApplied
 		results.applySuccess(gvk, nn)
 		tracker.isHealthy = isHealthy(lastApplied)
 		results.reportHealth(gvk, nn, tracker.isHealthy)
 	}
+
+	// We want to be more cautions on pruning and only do it if all manifests are applied.
+	if a.prune && results.applyFailCount == 0 {
+		klog.V(4).Infof("Prune is enabled")
+		err := func() error {
+			pruneObjects, err := kapplyset.FindAllObjectsToPrune(ctx, a.client, visitedUids)
+			if err != nil {
+				return err
+			}
+			if err = a.deleteObjects(ctx, pruneObjects, results); err != nil {
+				return err
+			}
+			return nil
+		}()
+		if err != nil {
+			klog.Errorf("encounter error on pruning %v", err)
+			if e := a.updateParentLabelsAndAnnotations(ctx, kapplyset, "superset"); e != nil {
+				return results, e
+			}
+		}
+	}
+	if err := a.updateParentLabelsAndAnnotations(ctx, kapplyset, "latest"); err != nil {
+		klog.Errorf("update parent failed %v", err)
+	}
 	return results, nil
 }
+
+func mergeMap(from, to map[string]string) map[string]string {
+	if to == nil {
+		to = make(map[string]string)
+	}
+	for k, v := range from {
+		to[k] = v
+	}
+	return to
+}
+
+// updateLabel adds the "applyset.kubernetes.io/part-of: Parent-ID" label to the manifest.
+func (a *ApplySet) updateManifestLabel(obj ApplyableObject, applysetLabels map[string]string) error {
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("unable to convert `ApplyableObject` to `unstructured.Unstructured` %v/%v %v",
+			obj.GetName(), obj.GetNamespace(), obj.GroupVersionKind().String())
+	}
+	labels := u.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	newLabels := mergeMap(applysetLabels, labels)
```

Comment:
mergeMap is nice, but it might be clearer just to inline it (e.g. here I think you're checking if `labels == nil` twice)

Reply:
Sure. But the mergeMap is called 3 times, the code could be a little bit repeating.